examples/md-flexible/src/ParticleSerializationTools.cpp

```diff
@@ -0,0 +1,74 @@
+/**
+ * @file ParticleSerializationTools.cpp
+ * @author J. Körner
+ * @date 13.05.2021
+ */
+#include "ParticleSerializationTools.h"
+
+namespace ParticleSerializationTools {
+void serializeParticle(ParticleType &particle, std::vector<char> &serializedParticles) {
+  ParticleAttributes attributes;
+
+  attributes.position = particle.getR();
+  attributes.velocity = particle.getV();
+  attributes.force = particle.getF();
+
+  if (particle.isDummy()) {
+    attributes.ownershipState = autopas::OwnershipState::dummy;
+  } else if (particle.isOwned()) {
+    attributes.ownershipState = autopas::OwnershipState::owned;
+  } else if (particle.isHalo()) {
+    attributes.ownershipState = autopas::OwnershipState::halo;
+  }
+
+  attributes.id = particle.getID();
+  attributes.typeId = particle.getTypeId();
+  attributes.oldForce = particle.getOldF();
+
+  std::vector<char> attributesVector;
+  attributesVector.resize(sizeof(ParticleAttributes));
+
+  std::memcpy(&attributesVector[0], &attributes, sizeof(ParticleAttributes));
+
+  ParticleType deserializedParticle;
+  deserializeParticleData(attributesVector.data(), deserializedParticle);
+
+  serializedParticles.insert(serializedParticles.end(), attributesVector.begin(), attributesVector.end());
```

Comment:
couldn't you make this simpler and at the same time eliminate the ParticleAttributes class? I'm thinking about something like this:

```cpp
template<class T>
void emplaceInBuffer(std::vector<unsigned char> buffer, T value) {
  const bufferSizeBefore = buffer.size();
  const sizeOfValue = sizeof(value);
  buffer.resize(bufferSizeBefore + sizeOfValue;
  std::memcpy(&buffer[bufferSizeBefore], &value, sizeOfValue);
}

// and then use like:
emplaceInBuffer(buff, particle.getR());
emplaceInBuffer(buff, particle.getV());
emplaceInBuffer(buff, particle.getID());

```
Mind typos etc, I'm tired ;)

Reply:
Does this really make it simpler? And I would also have to implement some function for deserialization, right? And without the ParticleAttributes the deserialization would get a lot messier. 

And always resizing the buffer for every single attribute seems like a lot of additional work.

Reply:
> Does this really make it simpler?
Debatable. I think it is more direct since you do not have the additional indirection with the `ParticleAttributes` object.

> And I would also have to implement some function for deserialization, right? 
Sure but this is analogous.

> And without the ParticleAttributes the deserialization would get a lot messier.
Hmm this way you would have to make sure to push / pop values in the correct order.

> And always resizing the buffer for every single attribute seems like a lot of additional work.
If you do one big `buffer.reserve()` beforehand the `resize()` calls are cheap.

I see merit in both ways. I do not like that we have to co-maintain the `ParticleAttributes`, however, paying attention to the order of pushing / popping values to the buffer is also weird. Probably a more elegant solution would be the push / pop approach but using the `MoleculeLJ::AttributeNames` and respective `get`. You would have to only define one tuple with all `AttributeNames` that should be sent and traverse it forward for serialization and backward for deserialization. There already is a function for iterating over a tuple in `TupleUtils.h`.

Reply:
I don't think the solution using AttributeNames is possible.
AttributeNames contains as first attribute the pointer to the object, which we don't want to send.
because the ParticleBase::get<>() function uses a template to identify the attribute, we need to define the getNeededAttribute as constexpr.
Therefore it is not possible to skip the pointer element in the Attribute names.

Additionally, because the current solution is working and because my thesis does not focus on finding the optimal way to serialize data for MPI I would mark this as a todo.
I dont' think it is reasonable to invest more time, here, even if ParticleAttributes needs to be maintained in the future.

Reply:
-> continued in PR #630